### PR TITLE
Switch to Ogmios for execution units evaluation

### PR DESCRIPTION
**Pull request by errfrom**
Closes https://github.com/Plutonomicon/cardano-transaction-lib/issues/174

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [x] **All** existing examples have been run locally against a fully-synced testnet node
- [x] The integration and unit tests have been run (`npm run test`) locally
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -32,7 +32,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - `Contract.Wallet.mkGeroWallet` and `Contract.Wallet.mkNamiWallet` - `Aff` versions should be used instead.
 - Protocol param update setters for the decentralisation constant (`set_d`) and the extra entropy (`set_extra_entropy`).
 - `AbsSlot` and related functions have been removed in favour of `Slot`.
-- Modules `Metadata.Seabug` and `Metadata.Seabug.Share`
+- Modules `Metadata.Seabug` and `Metadata.Seabug.Share`.
+- POST /eval-ex-units Haskell server endpoint.
```
Comment (klntsky) [inline CHANGELOG.md:36]:
```suggestion
- `POST /eval-ex-units` Haskell server endpoint. 
```

File: test/Ogmios/EvaluateTx.purs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+ogmiosEvaluateTxValidRespDecoded :: Array (RedeemerPointer /\ ExecutionUnits)
+ogmiosEvaluateTxValidRespDecoded =
+  [ { redeemerTag: Mint, redeemerIndex: nat zero }
+      /\ { memory: nat 1685698, steps: nat 609724445 }
+  , { redeemerTag: Spend, redeemerIndex: nat one }
+      /\ { memory: nat 1700, steps: nat 476468 }
+  ]
+
+nat :: Int -> Natural
```
Comment (klntsky) [inline test/Ogmios/EvaluateTx.purs:49]: Please rename it to `fromInt` and move to `Types.Natural`

**Review by klntsky — CHANGES_REQUESTED** [review]
Great, thank you!

**Review by ngua — COMMENTED** [review]
I haven't had the change to look at this very closely yet, and I can't leave a comment on the relevant part of `server/src/Types.hs`, but we can remove everything related to the node socket now, right? AFAIK we only needed that to perform queries related to the ex units calculation

File: src/Types/RedeemerTag.purs [hunk 3]
```diff
@@ -1,39 +1,28 @@
...
 -- lives in it's own module due to a name conflict with the `Mint` Type
 data RedeemerTag = Spend | Mint | Cert | Reward
 
+fromString :: String -> Maybe RedeemerTag
+fromString tag
+  | tag == "spend" = Just Spend
+  | tag == "mint" = Just Mint
+  | tag == "certificate" = Just Cert
+  | tag == "withdrawal" = Just Reward
+  | otherwise = Nothing
```
Comment (vvtran) [inline src/Types/RedeemerTag.purs:21]:
```
fromString :: String -> Maybe RedeemerTag
fromString = case _ of
  "spend" -> Just Spend
  "mint" -> Just Mint
  "certificate" -> Just Cert
  "withdrawal" -> Just Reward
  _ -> Nothing
```
might be nicer but down to you.

File: test/Ogmios/EvaluateTx.purs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+nat = Natural.fromBigInt' <<< BigInt.fromInt
```
Comment (vvtran) [inline test/Ogmios/EvaluateTx.purs:50]: I would also add to perhaps re-export this from `Contract.Numeric.Natural`

File: src/QueryM/Ogmios.purs [hunk 5]
```diff
@@ -441,21 +445,49 @@ instance Show SafeZone where
...
+    decodeRdmrPtrExUnitsItem (redeemerPtrRaw /\ exUnitsAeson) = do
+      redeemerPtr <- note redeemerPtrTypeMismatch $
+        decodeRedeemerPointer redeemerPtrRaw
+      flip aesonObject exUnitsAeson $ \exUnitsObj -> do
+        memory <- getField exUnitsObj "memory"
+        steps <- getField exUnitsObj "steps"
+        pure $ redeemerPtr /\ { memory, steps }
+
+    redeemerPtrTypeMismatch :: JsonDecodeError
+    redeemerPtrTypeMismatch = TypeMismatch $
```
Comment (vvtran) [inline src/QueryM/Ogmios.purs:479]: redundant `$`

**Review by vvtran — COMMENTED** [review]
Heads up @errfrom, I'm getting
```
Summary
2/16 tests passed
```
due to `(AtKey "coinsPerUtxoWord" MissingValue)`

Edit: ignore this

**Comment by vvtran** [comment]
I think there was a PR somewhere that updated the field name `coinsPerUtxoWord` no? @jy14898

**Comment by errfrom** [comment]
@vvtran Don't we have the changes related to the Babbage protocol parameters in `vasil` branch?

**Comment by errfrom** [comment]
@vvtran I believe we'll need to make some minor changes in `BalanceTx` as well because of the new `coinsPerUtxoByte`

**Comment by vvtran** [comment]
However, is this not a change we need to implement soon as our tests will continue to fail? A partial update happened yesterday right?

**Comment by vvtran** [comment]
I think I didn't switch nodes properly perhaps.

**Comment by vvtran** [comment]
Anyhow, I think 14/16 of the tests are passing on this branch anyway? So the posix time change is still required.

**Comment by vvtran** [comment]
@errfrom  Apologies, please ignore my comments, I got pretty confused because I've been switching between branches with different nodes and the tests were confusing me, The tests are indeed _passing_.
cc: @klntsky @ngua

**Review by klntsky — APPROVED** [review]
I approve this but it's broken without Ogmios fork update

**Comment by ngua** [comment]
I think we can close this since we need to merge `vasil` into `develop`, and these changes are already there

**Pull request closed without merging**